File: DFS/#110 Balanced Binary Tree.cpp

class Solution {
public:
    
    int depth(TreeNode* root){
        if(!root) return -1;
        return max(depth(root->left),depth(root->right))+1;
    }
   
    bool isBalanced(TreeNode* root) {
        if(!root) return true;
        if(!root->left && !root->right) return true;
        if(abs(depth(root->left) - depth(root->right))>1) return false;
        return isBalanced(root->left) && isBalanced(root->right);
    }
};

File: Tree/#110. Balanced Binary Tree.py
'''
TC: O(n)
SC: O(1)
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.is_balanced = True
        self.height(root)
        return self.is_balanced

    
    def height(self, root):
        if not root:
            return 0
        
        left_height = self.height(root.left)
        right_height = self.height(root.right)
        if abs(left_height - right_height) > 1:
            self.is_balanced = False

        return 1 + max(left_height, right_height)
        